### ISSUE TITLE
FCS-GX memory usage

### ISSUE THREAD

**Issue by fgvieira** [ncbi/fcs-gx#4]
How does `fcs-gx` uses memory? Does it load the entire DB into memory every time? If not, is there a way to estimate the approximate memory usage of `fcs-gx`?

According to the README file, `fcs-gx` needs ~450Gb of RAM. When running several thousand genomes, this can quickly become a constraint. Is there a way to reduce memory usage? Or maybe share the memory between different processes?

thanks,

**Comment by etvedte** [comment]
Hello,

I've copy-pasted some of your comments with my responses below:

> How does fcs-gx uses memory? Does it load the entire DB into memory every time?

Ideally gxdb is cached in shared memory so it is stable and doesn't need to be loaded for every run. If it is not cached, if consecutive GX calls occur in a short enough timeframe, the db files will persist

> If not, is there a way to estimate the approximate memory usage of fcs-gx?

Running `fcs.py` or `run_gx` debug mode `--debug` will provide details about memory usage

> According to the README file, fcs-gx needs ~450Gb of RAM. When running several thousand genomes, this can quickly become a constraint. Is there a way to reduce memory usage? Or maybe share the memory between different processes?

There is considerable slowdown on hosts with insufficient RAM. That being said, when running several thousand genomes on a host with sufficient RAM, the disk is only accessed once when the database is initially memory-mapped in RAM, and is subsequently shared by subsequent gx processes. Also, GX screens most genomes in a few minutes such that running hundreds or thousands of genomes is easily achieved.

Eric

**Comment by fgvieira** [comment]
@etvedte thanks for your quick reply!

Just to be sure I understand, when you say "gxdb is cached in shared memory", are you referring to what is mentioned in the README, that you copy the gxdb to (e.g.) `/dev/shm/gxdb`?

In that case all gx jobs running will use the same shared memory, being possible (in theory) to run several parallel gx jobs on a 512 GiB of RAM server, right?

And how do you think that would be reported on a SLURM system? Would the ~450Gb be reported for each job, or would it be considered "shared" and as such not counted for the jobs?

Lastly, is memory-mapping enabled by default or do I need to use a special option to enable it?

thanks,

**Comment by etvedte** [comment]
gxdb cached in shared memory refers to mounting it to tmpfs- or ramfs-backed filesystem as described here. It is enabled by default.
https://github.com/ncbi/fcs/wiki/FCS-GX-quickstart#download-the-fcs-gx-database

I'm not sure how it would be reported on SLURM. There are a few closed tickets with users running SLURM but I couldn't find any details on how they configured their jobs. 
https://github.com/ncbi/fcs/issues?q=SLURM

As stated above, you can run `--debug` to get memory reports for the tasks with and without parallelization, and then tailor to your needs. You can set cores with `export GX_NUM_CORES=N` when running from source code or set in an `env.txt` as described [here](https://github.com/ncbi/fcs/wiki/FCS-GX-input#environment-variables) if using the fcs.py runner. 

If you want to post your observations here it would be helpful for others running with similar schedulers. Thanks!

**Comment by fgvieira** [comment]
I tried running `fcs-gx` with gxdb on a tmpfs mount, and I was able to run 10 concurrent `fcs-gx` jobs on a 700Gb RAM server! :smile: 
This will greatly speed-up things!

I haven't been able to test it on SLURM yet.

**Comment by muffato** [comment]
> And how do you think that would be reported on a SLURM system? Would the ~450Gb be reported for each job, or would it be considered "shared" and as such not counted for the jobs?

Here, on LSF (not SLURM, I know, but I think it's still useful to mention), the ~400-500 GB of memory mapped database are _not_ considered by LSF and not included in the final job report. As far as LSF is concerned, a few GBs of RAM is all FCS-GX needs, and asking 512 GB RAM is only flagging the job as ~1% memory efficient to the sys-admins.
**However**, when asking only a few GBs RAM, we faced a good proportion of FCS-GX jobs being killed by the Linux kernel itself (Out Of Memory killer) because all the RAM was allocated to processes running on the machine. So in the end, on LSF, the best is to request the 512 GB advised in the docs to make sure that much RAM is free for the job.

In the tests above, I didn't check multiple FCS-GX jobs running on the same host, but I think the conclusion would still be that you need to make sure there's 512 GB RAM available to your jobs. I guess you could submit 2x 256 GB RAM jobs if you had the guarantee they would run on the same host at the same time. In our setup, given the wide range of users and jobs, I wouldn't take any chance :)

Hope this helps !

**Comment by murphyte** [comment]
> So in the end, on LSF, the best is to request the 512 GB advised in the docs to make sure that much RAM is free for the job.

yep, that's the crux of it. If you're screening *a lot*, then setting up a system where you can cluster jobs to the same server and request 512 GB for the first job but a small amount of memory for subsequent jobs is _ideal_, but not trivial. Actual FCS-GX memory usage can go up to I think maybe 20 GB in some cases, but the primary issue is to make sure there's enough memory to completely hold the index files _once_.

Another option is to make an additional wrapper that runs on your farm host and sequentially loops through a set of genomes, as a way to treat multiple genomes as a single farm job and ensure they all run on the same host. You could even use something like `xargs` to run them in parallel on that one host -- we used that approach for running GX across all existing genomes on a couple of hosts. You might set that wrapper up to be triggered from a cron, using a drop-box to queue up genomes to run and build up batches. But that approach is harder to fit into a larger pipeline, so it really depends on how many genomes you're screening and what your overhead is for caching the db into memory (read time).
